### PR TITLE
MPLUGIN-417 Descriptor with plain text by default

### PR DESCRIPTION
**Pull request by kwin**
Report uses enhanced descriptor with HTML
Include better links to javadoc
Include more block taglets in description
Fix plain text plugin descriptor to include all details and no longer include javadoc tags

### CODE REVIEW

**Comment by kwin** [comment]
@rfscholte What do you think of this approach and scanning the sources twice for plain-text and html separately?

**Comment by rfscholte** [comment]
The commit is too big for me to see what's happening. 
With TDD I'd like to see:
- plugin descriptor keeps the original javadoc, so inlcluding inline doclettags.
- the generated help-goal prints plain text
- the report generator of the goal transforms the inline doclettags to valid html.

IIRC in the past I've made a change that the two outputs are always based on the same source: the plugin descriptor.

**Comment by kwin** [comment]
> IIRC in the past I've made a change that the two outputs are always based on the same source: the plugin descriptor.

The whole point is that this approach prevents emitting the full details. Currently the plugin descriptor contains inline taglets (which is wrong) and the HTML report does no longer have access to the full context. So in order for the report to not go via

`(parse source file information -> convert javadoc to html -> convert html to plaintext -> serialize descriptor) -> (deserialize descriptor -> convert plain text to html)`
(while the former block being goal `descriptor` and the latter being `report`) you have to do the parsing twice.

**Comment by kwin** [comment]
> The commit is too big for me to see what's happening.

Right now you can completely ignore the newly added classes, just look at the modification of the existing ones, to tell me what you think of the approach in general.

**Comment by rfscholte** [comment]
I think I get your point regarding the context. Based on that it is indeed fine to use HTML as content for the plugin descriptor. Just make sure there's a test that requires this context, as it is the critical part of this solution.

File: maven-plugin-plugin/pom.xml [hunk 1]
```diff
@@ -319,6 +319,9 @@
                 <maven.compiler.source>${maven.compiler.source}</maven.compiler.source>
                 <maven.compiler.target>${maven.compiler.target}</maven.compiler.target>
               </properties>
+              <pomExcludes>
+                <pomExclude>plugin-info-jdk/pom.xml</pomExclude><!-- relies on just a given plugin.xml for generating a report, this is no longer supported -->
```
Comment (slawekjaranowski) [inline maven-plugin-plugin/pom.xml:323]: maybe remove at all

File: maven-plugin-tools-java/src/main/java/org/apache/maven/tools/plugin/extractor/javadoc/JavaJavadocMojoDescriptorExtractor.java [hunk 2]
```diff
@@ -56,11 +56,13 @@
 
 /**
  * <p>
```
Comment (kwin) [inline maven-plugin-tools-java/src/main/java/org/apache/maven/tools/plugin/extractor/javadoc/JavaJavadocMojoDescriptorExtractor.java:58]: Do we need to convert javadoc -> XHTML for the legacy extractor as well?

File: maven-script/maven-plugin-tools-beanshell/src/main/java/org/apache/maven/tools/plugin/extractor/beanshell/BeanshellMojoDescriptorExtractor.java [hunk 3]
```diff
@@ -131,6 +131,7 @@ private MojoDescriptor createMojoDescriptor( String basedir, String resource, Pl
             throw new InvalidPluginDescriptorException( "Error scanning beanshell script", evalError );
         }
 
+        // FIXME: convert javadocs
```
Comment (kwin) [inline maven-script/maven-plugin-tools-beanshell/src/main/java/org/apache/maven/tools/plugin/extractor/beanshell/BeanshellMojoDescriptorExtractor.java:134]: necessary to convert javadocs -> xhtml for deprecated BSH extractor?

**Review by slawekjaranowski — COMMENTED** [review]
Generally ok, a few question in code

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java [hunk 4]
```diff
@@ -88,44 +81,7 @@ public void execute()
...
-        if ( !"maven-plugin".equalsIgnoreCase( project.getArtifactId() )
-            && project.getArtifactId().toLowerCase().startsWith( "maven-" )
-            && project.getArtifactId().toLowerCase().endsWith( "-plugin" )
-            && !"org.apache.maven.plugins".equals( project.getGroupId() ) )
-        {
-            getLog().error( LS + LS + "Artifact Ids of the format maven-___-plugin are reserved for" + LS
-                                + "plugins in the Group Id org.apache.maven.plugins" + LS
-                                + "Please change your artifactId to the format ___-maven-plugin" + LS
-                                + "In the future this error will break the build." + LS + LS );
-        }
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:100]: Logging it more than once is a bit redundant and you always need to run descriptor after helpmojo anyways. Therefore just logging in descriptor seems to be enough

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java [hunk 5]
```diff
@@ -43,12 +42,6 @@
     @Parameter( defaultValue = "${project}", readonly = true )
     protected MavenProject project;
 
-    /**
-     * The goal prefix that will appear before the ":".
-     */
-    @Parameter
-    protected String goalPrefix;
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:50]: goalPrefix is almost useless in helpmojo. It is just used for the report on this mojo. Will try to remove this last remnant

File: maven-plugin-tools-java/src/main/java/org/apache/maven/tools/plugin/extractor/javadoc/JavaJavadocMojoDescriptorExtractor.java (hunk 2 above)
Comment (slawekjaranowski) [reply maven-plugin-tools-java/src/main/java/org/apache/maven/tools/plugin/extractor/javadoc/JavaJavadocMojoDescriptorExtractor.java:58]: I would not touch legacy extractors

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (slawekjaranowski) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:50]: It is need:
https://github.com/apache/maven-plugin-tools/blob/706b1d0b6730d028350f18d8459eee8b123e2f67/maven-plugin-tools-generators/src/main/resources/help-class-source.vm#L44

Without it HelpMojo.java will have unresolved value

[the remark this answers is not in the document]
Comment (kwin) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:50]: As I said: " is almost useless". This only affects the comments on the class which is evaluated by report. I would rather get rid of the prefix in the comment and and just reference the full Maven coordinates here. WDYT?

File: pom.xml [hunk 6]
```diff
@@ -94,9 +94,13 @@
     <javaVersion>8</javaVersion>
     <pluginTestingHarnessVersion>3.3.0</pluginTestingHarnessVersion>
     <mavenVersion>3.2.5</mavenVersion>
+    <!-- SLF4J version must match the version exported from the mavenVersion, 
+    https://github.com/apache/maven/blob/12a6b3acb947671f09b81f49094c53f426d8cea1/pom.xml#L63 -->
```
[the remark this answers is not in the document]
Comment (kwin) [reply pom.xml:98]: Not sure, as even minor versions may introduce new features which are backwards-compatible (i.e. new methods added to API). When you leverage those your code breaks when running with an older SLF4J API dependency.

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (slawekjaranowski) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:50]: It is needed for this change?
If no - can we move such change to separate PR?

[the remark this answers is not in the document]
Comment (kwin) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/AbstractGeneratorMojo.java:50]: I agree, will back out those unrelated changes.

File: maven-script/maven-plugin-tools-beanshell/src/main/java/org/apache/maven/tools/plugin/extractor/beanshell/BeanshellMojoDescriptorExtractor.java (hunk 3 above)
Comment (kwin) [reply maven-script/maven-plugin-tools-beanshell/src/main/java/org/apache/maven/tools/plugin/extractor/beanshell/BeanshellMojoDescriptorExtractor.java:151]: I left the legacy handling in place, i.e. convert inline javadoc tags afterwards to HTML (but only for legacy extractors).

**Review by gnodet — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #725

**Pull request merged**